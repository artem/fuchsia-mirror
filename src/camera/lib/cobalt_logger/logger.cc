// Copyright 2021 The Fuchsia Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "src/camera/lib/cobalt_logger/logger.h"

#include <lib/async/cpp/task.h>
#include <lib/syslog/cpp/macros.h>
#include <lib/zx/time.h>

#include <string>

#include "src/cobalt/bin/utils/error_utils.h"
#include "src/lib/fxl/strings/string_printf.h"

namespace camera::cobalt {
namespace {

using async::PostDelayedTask;
using ::cobalt::ErrorToString;
using fuchsia::metrics::MetricEventLoggerFactory;
using fxl::StringPrintf;

// Keep this under 1750 so we can't overflow the cobalt channel.
constexpr uint32_t kMaxPendingEvents = 1700u;

uint64_t CurrentTimeUSecs(const std::unique_ptr<timekeeper::Clock>& clock) {
  return zx::nsec(clock->Now().get()).to_usecs();
}

inline bool HasStreamProperty(fuchsia::camera2::CameraStreamType stream_type,
                              fuchsia::camera2::CameraStreamType target_type) {
  return (stream_type & target_type) == target_type;
}

}  // namespace

Logger::Logger(async_dispatcher_t* dispatcher, std::shared_ptr<sys::ServiceDirectory> services,
               std::unique_ptr<timekeeper::Clock> clock)
    : dispatcher_(dispatcher),
      services_(services),
      clock_(std::move(clock)),
      logger_reconnection_backoff_(/*initial_delay=*/zx::msec(100), /*retry_factor=*/2u,
                                   /*max_delay=*/zx::hour(1)) {
  logger_.set_error_handler([this](zx_status_t status) {
    FX_PLOGS(WARNING, status) << "Lost connection with MetricEventLogger";
    RetryConnectingToLogger();
  });

  auto logger_request = logger_.NewRequest();
  ConnectToLogger(std::move(logger_request));
}

void Logger::Shutdown() {
  shut_down_ = true;

  pending_events_.clear();
  timer_starts_usecs_.clear();

  reconnect_task_.Cancel();

  logger_factory_.Unbind();
  logger_.Unbind();
}

void Logger::ConnectToLogger(
    ::fidl::InterfaceRequest<fuchsia::metrics::MetricEventLogger> logger_request) {
  // Connect to the LoggerFactory.
  logger_factory_ = services_->Connect<fuchsia::metrics::MetricEventLoggerFactory>();

  logger_factory_.set_error_handler([](zx_status_t status) {
    FX_PLOGS(WARNING, status) << "Lost connection with MetricEventLoggerFactory";
  });

  fuchsia::metrics::ProjectSpec project;
  project.set_customer_id(kCustomerId);
  project.set_project_id(kProjectId);

  logger_factory_->CreateMetricEventLogger(
      std::move(project), std::move(logger_request),
      [this](fuchsia::metrics::MetricEventLoggerFactory_CreateMetricEventLogger_Result result) {
        // We don't need a long standing connection to the LoggerFactory so we unbind after
        // setting up the Logger.
        logger_factory_.Unbind();

        if (result.is_response()) {
          logger_reconnection_backoff_.Reset();
        } else if (result.err() == fuchsia::metrics::Error::SHUT_DOWN) {
          FX_LOGS(INFO) << "Stopping sending Cobalt events";
          logger_.Unbind();
        } else {
          FX_LOGS(WARNING) << "Failed to set up Cobalt: " << ErrorToString(result.err());
          logger_.Unbind();
          RetryConnectingToLogger();
        }
      });
}

void Logger::RetryConnectingToLogger() {
  if (logger_) {
    return;
  }

  // Bind |logger_| and immediately send the events that were not acknowledged by the server on the
  // previous connection.
  auto logger_request = logger_.NewRequest();
  SendAllPendingEvents();

  reconnect_task_.Reset([this, request = std::move(logger_request)]() mutable {
    ConnectToLogger(std::move(request));
  });

  PostDelayedTask(
      dispatcher_, [reconnect = reconnect_task_.callback()]() { reconnect(); },
      logger_reconnection_backoff_.GetNext());
}

void Logger::LogEvent(Event event) {
  FX_CHECK(!shut_down_);
  if (pending_events_.size() >= kMaxPendingEvents) {
    // Drop event
    dropped_events_since_last_report_++;
    auto now = zx::clock::get_monotonic();
    if (now < droped_events_next_report_permitted_) {
      // Avoid logging to because not enough time has passed since the last report.
      return;
    }
    FX_LOGS(WARNING) << StringPrintf(
        "Dropped %lu Cobalt events (event: %s) - too many pending events (%lu)",
        dropped_events_since_last_report_, event.ToString().c_str(), pending_events_.size());
    dropped_events_since_last_report_ = 0;
    droped_events_next_report_permitted_ = now + zx::sec(1);
    return;
  }

  const uint64_t event_id = next_event_id_++;
  pending_events_.emplace(event_id, std::move(event));
  SendEvent(event_id);
}

uint64_t Logger::StartTimer() {
  FX_CHECK(!shut_down_);

  const uint64_t timer_id = next_event_id_++;
  timer_starts_usecs_.insert(std::make_pair(timer_id, CurrentTimeUSecs(clock_)));
  return timer_id;
}

void Logger::SendEvent(uint64_t event_id) {
  if (!logger_) {
    return;
  }

  if (pending_events_.find(event_id) == pending_events_.end()) {
    return;
  }
  Event& event = pending_events_.at(event_id);

  auto callback = [this, event_id,
                   &event](fuchsia::metrics::MetricEventLogger_LogOccurrence_Result result) {
    if (result.is_err()) {
      FX_LOGS(ERROR) << StringPrintf("Cobalt logging error: status %s, event %s",
                                     ErrorToString(result.err()).c_str(), event.ToString().c_str());
    }

    // We don't retry events that have been acknowledged by the server, regardless of the return
    // status.
    pending_events_.erase(event_id);
  };

  switch (event.GetType()) {
    case EventType::kOccurrence:
      logger_->LogOccurrence(event.GetMetricId(), 1u, event.GetDimensions(), std::move(callback));
      break;

    default:
      FX_LOGS(WARNING) << "Ignores an event with unknown type.";
      break;
  }
}

void Logger::SendAllPendingEvents() {
  for (const auto& [event_id, _] : pending_events_) {
    SendEvent(event_id);
  }
}

uint64_t Logger::GetTimerDurationUSecs(uint64_t timer_id) const {
  FX_CHECK(timer_starts_usecs_.find(timer_id) != timer_starts_usecs_.end());

  return CurrentTimeUSecs(clock_) - timer_starts_usecs_.at(timer_id);
}

StreamType Logger::ConvertStreamType(fuchsia::camera2::CameraStreamType type) {
  if (HasStreamProperty(type, fuchsia::camera2::CameraStreamType::VIDEO_CONFERENCE)) {
    if (HasStreamProperty(type, fuchsia::camera2::CameraStreamType::EXTENDED_FOV)) {
      if (HasStreamProperty(type, fuchsia::camera2::CameraStreamType::FULL_RESOLUTION)) {
        return cobalt::StreamType::kStream5;
      } else {
        return cobalt::StreamType::kStream6;
      }
    } else {
      if (HasStreamProperty(type, fuchsia::camera2::CameraStreamType::FULL_RESOLUTION)) {
        return cobalt::StreamType::kStream3;
      } else {
        return cobalt::StreamType::kStream4;
      }
    }
  } else {
    if (HasStreamProperty(type, fuchsia::camera2::CameraStreamType::MONITORING)) {
      return cobalt::StreamType::kStream2;
    } else if (HasStreamProperty(type, fuchsia::camera2::CameraStreamType::FULL_RESOLUTION)) {
      return cobalt::StreamType::kStream0;
    } else {
      return cobalt::StreamType::kStream1;
    }
  }
}

}  // namespace camera::cobalt
