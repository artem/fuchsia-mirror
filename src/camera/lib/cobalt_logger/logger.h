// Copyright 2021 The Fuchsia Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#ifndef SRC_CAMERA_LIB_COBALT_LOGGER_LOGGER_H_
#define SRC_CAMERA_LIB_COBALT_LOGGER_LOGGER_H_

#include <fuchsia/camera2/cpp/fidl.h>
#include <fuchsia/metrics/cpp/fidl.h>
#include <lib/fit/function.h>
#include <lib/sys/cpp/service_directory.h>
#include <lib/zx/time.h>

#include <any>
#include <map>

#include "src/camera/lib/cobalt_logger/event.h"
#include "src/lib/backoff/exponential_backoff.h"
#include "src/lib/fxl/functional/cancelable_callback.h"
#include "src/lib/timekeeper/clock.h"
#include "src/lib/timekeeper/system_clock.h"

namespace camera::cobalt {

template <typename T>
constexpr auto ToIntegral(T t) -> typename std::underlying_type<T>::type {
  return static_cast<typename std::underlying_type<T>::type>(t);
}

// Log events to cobalt.
class Logger {
 public:
  // We expect fuchsia.metrics.MetricEventLoggerFactory to be in |services|.
  Logger(async_dispatcher_t* dispatcher, std::shared_ptr<sys::ServiceDirectory> services,
         std::unique_ptr<timekeeper::Clock> clock = std::make_unique<timekeeper::SystemClock>());

  virtual ~Logger() = default;

  // Log event with no dimensions.
  virtual void LogInteger(uint32_t metric_id, std::vector<uint32_t> dimensions, uint64_t value) {
    LogEvent(CameraEvent(EventType::kInteger, metric_id, std::move(dimensions), value));
  }

  // Log event with fuchsia.metrics.MetricEventLogger with the provided parameters. If the service
  // is not accessible, keep the parameters to try again later.
  virtual void LogOccurrence(uint32_t metric_id, std::vector<uint32_t> dimensions) {
    LogEvent(CameraEvent(EventType::kOccurrence, metric_id, std::move(dimensions), 1lu));
  }

  // Start a timer and return the id to that timer. The id is needed to log the elapsed time since
  // starting the timer.
  uint64_t StartTimer();

  // Log the time elapsed in microseconds since starting the timer with id |timer_id| with
  // fuchsia.metrics.MetricEventLogger. If the service is not accessible, keep the parameters to
  // try again later.
  //
  // This does not stop the timer.
  virtual void LogElapsedTime(uint32_t metric_id, std::vector<uint32_t> dimensions,
                              uint64_t timer_id) {
    LogEvent(CameraEvent(EventType::kInteger, metric_id, std::move(dimensions),
                         GetTimerDurationUSecs(timer_id)));
  }

  // Log a duration in microseconds. If the service is not accessible, keep the parameters to try
  // again later.
  virtual void LogDuration(uint32_t metric_id, std::vector<uint32_t> dimensions,
                           zx::duration duration) {
    LogEvent(
        CameraEvent(EventType::kInteger, metric_id, std::move(dimensions), duration.to_usecs()));
  }

  // Immediately shutdown |Logger| so it can no longer be used to log events.
  void Shutdown();

  template <typename DimensionType, typename... RestDimensionTypes>
  static std::vector<uint32_t> BuildDimension(DimensionType dimension, RestDimensionTypes... rest) {
    std::vector<uint32_t> output;
    AddDimensions(output, dimension, rest...);
    return output;
  }

  static StreamType ConvertStreamType(fuchsia::camera2::CameraStreamType type);

 protected:
  virtual void LogEvent(Event event);

 private:
  template <typename DimensionType>
  static void AddDimensions(std::vector<uint32_t>& output, DimensionType dimension) {
    output.push_back(ToIntegral(dimension));
  }

  template <typename DimensionType, typename... RestDimensionTypes>
  static void AddDimensions(std::vector<uint32_t>& output, DimensionType dimension,
                            RestDimensionTypes... rest) {
    output.push_back(ToIntegral(dimension));
    AddDimensions(output, rest...);
  }

  void ConnectToLogger(
      ::fidl::InterfaceRequest<fuchsia::metrics::MetricEventLogger> logger_request);
  void RetryConnectingToLogger();
  void SendEvent(uint64_t event_id);
  void SendAllPendingEvents();
  uint64_t GetTimerDurationUSecs(uint64_t timer_id) const;

  async_dispatcher_t* dispatcher_;
  std::shared_ptr<sys::ServiceDirectory> services_;
  std::unique_ptr<timekeeper::Clock> clock_;

  fuchsia::metrics::MetricEventLoggerFactoryPtr logger_factory_;
  fuchsia::metrics::MetricEventLoggerPtr logger_;

  // An event is pending if it has been written into a channel, but has not been acknowledged by
  // the recipient.
  std::map<uint64_t, Event> pending_events_;
  std::map<uint64_t, uint64_t> timer_starts_usecs_;
  backoff::ExponentialBackoff logger_reconnection_backoff_;

  // We need to be able to cancel a posted reconnection task when |Logger| is destroyed.
  fxl::CancelableClosure reconnect_task_;

  // Batch up dropped event log messages, to prevent flooding the logs
  uint64_t dropped_events_since_last_report_ = 0;
  zx::time droped_events_next_report_permitted_;

  uint64_t next_event_id_ = 0;
  bool shut_down_ = false;
};

}  // namespace camera::cobalt

#endif  // SRC_CAMERA_LIB_COBALT_LOGGER_LOGGER_H_
