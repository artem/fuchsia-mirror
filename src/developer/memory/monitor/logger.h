// Copyright 2021 The Fuchsia Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#ifndef SRC_DEVELOPER_MEMORY_MONITOR_LOGGER_H_
#define SRC_DEVELOPER_MEMORY_MONITOR_LOGGER_H_

#include <lib/async/cpp/task.h>
#include <lib/async/dispatcher.h>

#include "src/developer/memory/metrics/capture.h"
#include "src/developer/memory/metrics/digest.h"
#include "src/developer/memory/monitor/memory_monitor_config.h"
#include "src/developer/memory/monitor/pressure_observer.h"
#include "src/lib/fxl/macros.h"

namespace monitor {

class Logger {
 public:
  using CaptureCb = fit::function<zx_status_t(memory::Capture*)>;
  using DigestCb = fit::function<void(const memory::Capture&, memory::Digest*)>;

  Logger(async_dispatcher_t* dispatcher, CaptureCb capture_cb, DigestCb digest_cb,
         memory_monitor_config::Config* config)
      : dispatcher_(dispatcher),
        capture_cb_(std::move(capture_cb)),
        digest_cb_(std::move(digest_cb)),
        config_(config) {}

  void SetPressureLevel(Level l);

 private:
  void Log();
  async_dispatcher_t* dispatcher_;
  CaptureCb capture_cb_;
  DigestCb digest_cb_;
  memory_monitor_config::Config* config_;
  zx::duration duration_;
  async::TaskClosureMethod<Logger, &Logger::Log> task_{this};

  FXL_DISALLOW_COPY_AND_ASSIGN(Logger);
};

}  // namespace monitor

#endif  // SRC_DEVELOPER_MEMORY_MONITOR_LOGGER_H_
