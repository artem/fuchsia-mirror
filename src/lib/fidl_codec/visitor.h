// Copyright 2019 The Fuchsia Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#ifndef SRC_LIB_FIDL_CODEC_VISITOR_H_
#define SRC_LIB_FIDL_CODEC_VISITOR_H_

#include "src/lib/fidl_codec/wire_object.h"
#include "src/lib/fidl_codec/wire_types.h"

namespace fidl_codec {

// Superclass for implementing visitors for Values. Note that the whole class is protected. To use a
// visitor, use the Visit method on the Value object you want to visit.
class Visitor {
 protected:
  virtual void VisitValue(const Value* node, const Type* for_type) {}
  virtual void VisitInvalidValue(const InvalidValue* node, const Type* for_type) {
    VisitValue(node, for_type);
  }
  virtual void VisitEmptyPayloadValue(const EmptyPayloadValue* node, const Type* for_type) {
    VisitValue(node, for_type);
  }
  virtual void VisitNullValue(const NullValue* node, const Type* for_type) {
    VisitValue(node, for_type);
  }
  virtual void VisitRawValue(const RawValue* node, const Type* for_type) {
    VisitValue(node, for_type);
  }
  virtual void VisitBoolValue(const BoolValue* node, const Type* for_type) {
    VisitValue(node, for_type);
  }
  virtual void VisitIntegerValue(const IntegerValue* node, const Type* for_type) {
    VisitValue(node, for_type);
  }
  virtual void VisitActualAndRequestedValue(const ActualAndRequestedValue* node,
                                            const Type* for_type) {
    VisitValue(node, for_type);
  }
  virtual void VisitDoubleValue(const DoubleValue* node, const Type* for_type) {
    VisitValue(node, for_type);
  }
  virtual void VisitStringValue(const StringValue* node, const Type* for_type) {
    VisitValue(node, for_type);
  }
  virtual void VisitHandleValue(const HandleValue* node, const Type* for_type) {
    VisitValue(node, for_type);
  }
  virtual void VisitUnionValue(const UnionValue* node, const Type* for_type) {
    node->value()->Visit(this, node->member().type());
  }
  virtual void VisitStructValue(const StructValue* node, const Type* for_type) {
    for (const auto& field : node->fields()) {
      field.second->Visit(this, field.first->type());
    }
  }
  virtual void VisitVectorValue(const VectorValue* node, const Type* for_type) {
    FX_DCHECK(for_type != nullptr);
    const Type* component_type = for_type->GetComponentType();
    FX_DCHECK(component_type != nullptr);
    for (const auto& value : node->values()) {
      value->Visit(this, component_type);
    }
  }
  virtual void VisitTableValue(const TableValue* node, const Type* for_type) {
    for (const auto& member : node->members()) {
      member.second->Visit(this, member.first->type());
    }
  }
  virtual void VisitFidlMessageValue(const FidlMessageValue* node, const Type* for_type) {
    if (node->decoded_request() != nullptr) {
      node->decoded_request()->Visit(this, nullptr);
    }
    if (node->decoded_response() != nullptr) {
      node->decoded_response()->Visit(this, nullptr);
    }
  }

  friend class Value;
  friend class InvalidValue;
  friend class EmptyPayloadValue;
  friend class NullValue;
  friend class RawValue;
  friend class BoolValue;
  friend class IntegerValue;
  friend class ActualAndRequestedValue;
  friend class DoubleValue;
  friend class StringValue;
  friend class HandleValue;
  friend class UnionValue;
  friend class StructValue;
  friend class VectorValue;
  friend class TableValue;
  friend class FidlMessageValue;
};

}  // namespace fidl_codec

#endif  // SRC_LIB_FIDL_CODEC_VISITOR_H_
