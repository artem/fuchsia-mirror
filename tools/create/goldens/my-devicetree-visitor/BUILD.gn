# Copyright 2020 The Fuchsia Authors. All rights reserved.
# Use of this source code is governed by a BSD-style license that can be
# found in the LICENSE file.

# This file was generated by the `fx create` command. The template
# is located at `//tools/create/templates/devicetree-visitor/BUILD.gn.tmpl-cpp`.
# If you find something broken, we are eager to review fixes.

import("//build/components.gni")
import("//build/devicetree/devicetree.gni")
import("//build/test.gni")

group("tests") {
  deps = [ ":my-devicetree-visitor-test" ]
  testonly = true
}

devicetree_visitor("my-devicetree-visitor") {
  deps = [ ":my-devicetree-visitor-impl" ]
}

source_set("my-devicetree-visitor-impl") {
  sources = [
    "my-devicetree-visitor.cc",
    "my-devicetree-visitor.h",
  ]
  deps = [
    "//sdk/lib/driver/devicetree/manager",
    "//sdk/lib/driver/devicetree/visitors:registration",
    "//sdk/lib/driver/logging/cpp",
  ]
}

test("my-devicetree-visitor-test-bin") {
  sources = [ "test/my-devicetree-visitor-test.cc" ]
  deps = [
    ":my-devicetree-visitor-impl",
    "//sdk/lib/driver/devicetree/testing:visitor-test-helper",
    "//sdk/lib/driver/devicetree/visitors:registry",
    "//sdk/lib/driver/devicetree/visitors/default/bind-property",
    "//src/lib/fxl/test:gtest_main",
  ]
}

dtb("my-devicetree-visitor-dtb") {
  sources = [ "test/dts/my-devicetree-visitor.dts" ]
}

resource("my-devicetree-visitor-dtb-file") {
  sources = get_target_outputs(":my-devicetree-visitor-dtb")
  outputs = [ "test-data/my-devicetree-visitor.dtb" ]
  deps = [ ":my-devicetree-visitor-dtb" ]
}

fuchsia_unittest_package("my-devicetree-visitor-test") {
  component_name = "my-devicetree-visitor-test"
  deps = [
    ":my-devicetree-visitor-dtb-file",
    ":my-devicetree-visitor-test-bin",
  ]
}
