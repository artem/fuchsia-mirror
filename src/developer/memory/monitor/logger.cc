// Copyright 2021 The Fuchsia Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "src/developer/memory/monitor/logger.h"

#include <lib/syslog/cpp/macros.h>

#include "lib/zx/time.h"
#include "src/developer/memory/metrics/printer.h"

namespace monitor {

void Logger::SetPressureLevel(Level l) {
  switch (l) {
    case kImminentOOM:
      duration_ = zx::sec(config_->imminent_oom_capture_delay_s());
      break;
    case kCritical:
      duration_ = zx::sec(config_->critical_capture_delay_s());
      break;
    case kWarning:
      duration_ = zx::sec(config_->warning_capture_delay_s());
      break;
    case kNormal:
      duration_ = zx::sec(config_->normal_capture_delay_s());
      break;
    case kNumLevels:
      break;
  }
  if (config_->capture_on_pressure_change()) {
    task_.Cancel();
    task_.PostDelayed(dispatcher_, zx::usec(1));
  }
}

void Logger::Log() {
  memory::Capture c;
  auto s = capture_cb_(&c);
  if (s != ZX_OK) {
    FX_LOGS_FIRST_N(INFO, 1) << "Error getting Capture: " << s;
    return;
  }
  memory::Digest d;
  digest_cb_(c, &d);
  std::ostringstream oss;
  memory::Printer p(oss);

  p.PrintDigest(d);
  auto str = oss.str();
  std::replace(str.begin(), str.end(), '\n', ' ');
  FX_LOGS(INFO) << str;

  task_.PostDelayed(dispatcher_, duration_);
}

}  // namespace monitor
