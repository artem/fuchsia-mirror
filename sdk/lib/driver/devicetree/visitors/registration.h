// Copyright 2023 The Fuchsia Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#ifndef LIB_DRIVER_DEVICETREE_VISITORS_REGISTRATION_H_
#define LIB_DRIVER_DEVICETREE_VISITORS_REGISTRATION_H_

#include <lib/driver/devicetree/manager/visitor.h>
#include <lib/driver/logging/cpp/logger.h>

#include <memory>

// The |VisitorRegistration| is the ABI for visitors to expose themselves to
// the Devicetree Visitor Framework. The visitor is loaded in as a shared
// library (also referred to as a DSO), and the global symbol
// `__devicetree_visitor_registration__` is used by the Devicetree Visitor
// Framework to locate this VisitorRegistration in the visitor library. The
// framework will use this to instantiate a visitor object to parse the
// devicetree.
struct VisitorRegistration {
  // The version is used in case we evolve the initialization signature
  // for any reason.
  uint64_t version;

  struct v1 {
    // Return a unique pointer to the visitor object or empty object in case of
    // error.
    // TODO(fxbug.dev/134108) : Temporarily using C++ ABI. Move to C ABI once
    // visitor interface is stabilized.
    std::unique_ptr<fdf_devicetree::Visitor> (*create_visitor)(fdf::Logger* logger);
  } v1;
};

#define VISITOR_REGISTRATION_VERSION_1 1

#define VISITOR_REGISTRATION_VERSION_MAX VISITOR_REGISTRATION_VERSION_1

#define DEVICETREE_VISITOR_REGISTRATION_V1(create_visitor) \
  { .version = VISITOR_REGISTRATION_VERSION_1, .v1 = {create_visitor}, }

#define EXPORT_DEVICETREE_VISITOR_REGISTRATION_V1(create_visitor)                   \
  extern "C" const VisitorRegistration __devicetree_visitor_registration__ __EXPORT \
  DEVICETREE_VISITOR_REGISTRATION_V1(create_visitor)

#define REGISTER_DEVICETREE_VISITOR(visitor_class)                          \
  EXPORT_DEVICETREE_VISITOR_REGISTRATION_V1(                                \
      [](fdf::Logger* logger) -> std::unique_ptr<fdf_devicetree::Visitor> { \
        fdf::Logger::SetGlobalInstance(logger);                             \
        return std::make_unique<visitor_class>();                           \
      })

#endif  // LIB_DRIVER_DEVICETREE_VISITORS_REGISTRATION_H_
