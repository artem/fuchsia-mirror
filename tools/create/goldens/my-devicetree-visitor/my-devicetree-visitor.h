// Copyright 2020 The Fuchsia Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#ifndef TOOLS_CREATE_GOLDENS_MY_DEVICETREE_VISITOR_MY_DEVICETREE_VISITOR_H_
#define TOOLS_CREATE_GOLDENS_MY_DEVICETREE_VISITOR_MY_DEVICETREE_VISITOR_H_


#include <lib/driver/devicetree/manager/visitor.h>

namespace my_devicetree_visitor_dt {

class MyDevicetreeVisitor : public fdf_devicetree::Visitor {
 public:
  MyDevicetreeVisitor() = default;
  zx::result<> Visit(fdf_devicetree::Node& node,
                     const devicetree::PropertyDecoder& decoder) override;
};

}  // namespace my_devicetree_visitor_dt

#endif  // TOOLS_CREATE_GOLDENS_MY_DEVICETREE_VISITOR_MY_DEVICETREE_VISITOR_H_
