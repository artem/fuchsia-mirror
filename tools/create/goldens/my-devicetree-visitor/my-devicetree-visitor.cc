// Copyright 2020 The Fuchsia Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "my-devicetree-visitor.h"

#include <lib/driver/devicetree/visitors/registration.h>
#include <lib/driver/logging/cpp/logger.h>

namespace my_devicetree_visitor_dt {

zx::result<> MyDevicetreeVisitor::Visit(fdf_devicetree::Node& node,
                                   const devicetree::PropertyDecoder& decoder) {
  return zx::ok();
}

}  // namespace my_devicetree_visitor_dt

REGISTER_DEVICETREE_VISITOR(my_devicetree_visitor_dt::MyDevicetreeVisitor);
