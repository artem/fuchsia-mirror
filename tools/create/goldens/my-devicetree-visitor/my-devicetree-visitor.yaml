# Copyright 2020 The Fuchsia Authors. All rights reserved.
# Use of this source code is governed by a BSD-style license that can be
# found in the LICENSE file.

%YAML 1.2
---
$id: http://devicetree.org/schemas/my-devicetree-visitor.yaml#
$schema: http://devicetree.org/meta-schemas/core.yaml#

title:  <Please add a title for this binding>

maintainers:
  - <Author name (author_email)>

description: |
  <Please add description for this binding>

properties:

examples:
  - |
    <Please add an example devicetree>
